### DISCUSSION TITLE
Improve and fix FileUpload acceptedFileTypes validation

### DISCUSSION THREAD

**Discussion by cododel** [filamentphp/filament#13254] [General]
Inconsistency between file validation logic between browser and laravel

For example, when uploading .js file with the mimetype text/javascript in acceptedFileTypes - the frontend successfully receives and saves the file to temporary storage

And when submitting a form, the backend already throws an error if the mimetype text/plain is missing

Fair and vice versa
The frontend will not accept a file without text/javascript

**Comment by cododel** [comment]
I still don't think this is a filepond problem
The documentation says that it checks the file type in two steps: by setting the basic html accept attribute on the html5 input element
https://pqina.nl/filepond/docs/api/plugins/file-validate-type/
<img width="506" alt="1" src="https://github.com/filamentphp/filament/assets/117383034/4bdea5c2-a93c-4a61-85e3-89f097326246">

This way the available files are filtered in the file manager
For example, with `['text/javascript']` - only files with the .js extension will be available
And with `['text/javascript', 'text/plain']` - csv files will also be available, since the browser also considers them files of the text/plain type

Further it is written in the documentation that if the browser does not correctly determine the file type, it additionally double-checks the type of the selected file based on its contents
This explains why with `['text/javascript', 'text/plain']` and a csv file selected, an error appears regarding the file type even before it is loaded
<img width="610" alt="2" src="https://github.com/filamentphp/filament/assets/117383034/58bf8abc-4705-42cb-8e22-90e986ef2054">

I downloaded the filepond source code, found an implementation for checking the file type based on its contents, and made a small testbed on playcode.io
https://playcode.io/1905242
<img width="818" alt="3" src="https://github.com/filamentphp/filament/assets/117383034/9da7d524-6b0a-411e-9ea3-50d9070bd60e">
Using FileReader, it reads the contents of the file in DataURI format, which at the beginning contains the mimetype that was obtained based on the contents of the file
And at this stage, the received type is compared with the passed list for validation, and the resulting text/csv is missing in `['text/javascript', 'text/plain']`, as a result an error occurs even before loading into the temporary livewire directory begins
<img width="1220" alt="4" src="https://github.com/filamentphp/filament/assets/117383034/5972c189-cab5-4e60-b8e6-61bf13c96e84">
<img width="506" alt="5" src="https://github.com/filamentphp/filament/assets/117383034/55cf368d-67a8-49b1-b44b-eac09141700d">
<img width="787" alt="6" src="https://github.com/filamentphp/filament/assets/117383034/0833d7fe-247a-4fae-85e6-fda3df614523">

But at the same time, if only `['text/javascript']` is specified in the validation, then the browser will allow you to select a js file, and it will pass content validation, as a result of which the download to the temporary livewire directory begins.
But already on the livewire side the file will be saved with an extension corresponding to its mimetype, which is determined by laravel
The reason for this is described here: [https://securinglaravel.com/laravel-security-file-upload-vulnerability/#1-don’t-trust-user-input](https://securinglaravel.com/laravel-security-file-upload-vulnerability/#1-don%E2%80%99t-trust-user-input)
The type is defined as text/plain
I assume with 85% confidence that the check occurs using the "file" system utility, using the command `file -b --mime index.js`
Which produces text/plain
But validation does not occur at this stage, since the file is loaded into a private temporary storage, in which, under normal conditions, it cannot be executed by anyone

When the loading is complete, the submit button for the form becomes available.
When you click it, laravel will check the mimetype according to the specified rule
Then file -b --mime index.js will output text/plain, but it will not match ['text/javascript'], and will display an error from the backend, under the FileUpload field widget

This is where the inconsistency lies not in libraries or plugins, but in the logic of the browser and the operating system.
And it is filementphp that transmits the specified rules to filepond and laravel, which use them during validation
Accordingly, for different logics, it should be possible to differentiate the rules
I believe that by default everything should be as is, and from acceptedFileTypes the rules should be passed to laravel and filepond "as is", without changes
But the acceptedBrowserFileTypes method must be added, which will overwrite the rules passed to filepond
In this case, I will write the following code:
```php
FileUpload::make('file')
 ->acceptedBrowserFileTypes(['text/javascript'])
 ->acceptedFileTypes(['text/plain'])
 ->getUploadedFileNameForStorageUsing(
 fn(TemporaryUploadedFile $file): string => (string) Str::uuid() . '.' . Arr::last(explode('.', $file->getClientOriginalName())),
 );
```

At first, the browser will allow you to select only files with the .js extension
Then filepond will check if the mimetype of the file specified in its extension matches the passed `['text/javascript']`
Next, the file will be uploaded to the temporary livewire directory with the extension .txt
Then, when submitting the form, the file will be double-checked by mimetype by Laravel itself using `['text/plain']` and make sure once again that it is a text file
Then it will be renamed, in accordance with the callback I passed to getUploadedFileNameForStorageUsing
And will be saved with its real extension 
The same is true for a large mass of other file types, I wrote about them above, and over the past two days I have also seen them in many discussions, there were .crt, .csv, .pen, .p12, .js and many other files
And such files are defined not only in the text/plain direction, but also binary ones, and media files sometimes do not match the mimetype on the frontend and backend
This is a kind of bug report it worked out

I think I was right in inadvertently thinking about the possibility of separating the rules for the frontend and backend
But didn't describe it in enough detail 
I'll downloading source code of filemantphp to research posiibility of this changes


anyway laravel and browser define mime type differently
https://youtu.be/PiQvvIbSUeI

**Comment by cododel** [comment]
filamentphp sends validation rules directly to filepond, which are passed to acceptedFileTypes "as is"

And it has the ability to transfer individual rules to it, duplicating the logic of a separate method and saving it in a separate attribute
And pass the rules to filepond in the form `acceptedBrowserFileTypes ?? acceptedFileTypes`

<img width="1552" alt="1" src="https://github.com/filamentphp/filament/assets/117383034/0fca1710-5707-4052-9af3-bf847a429b9e">
<img width="1552" alt="2" src="https://github.com/filamentphp/filament/assets/117383034/0789bcc5-a12f-4c8b-9bfa-ad70b02eb572">
<img width="1552" alt="3" src="https://github.com/filamentphp/filament/assets/117383034/da889654-4f80-4a0a-aa86-dfb9cee183a0">
<img width="1552" alt="4" src="https://github.com/filamentphp/filament/assets/117383034/9004f88f-7be6-4b4e-b912-23065a259813">


but in the filament source code itself, a solution is used that enumerates mime types, among which there is also text/plain for csv

Although for csv in general this is quite understandable, since the file can contain one word or a space delimiter

On the other hand, I’m a little distracted from the question and now I’m thinking that a js file can be defined as text/plain by mimetype, in case of obfuscation

For example, I have already made an obfuscator that cannot be decrypted by available online decryptors, then you can only identify it by extension, blindly trusting
In this case, this could be true, but the browser will consider such a file valid if it has a .js extension
I'll check this hypothesis now

I think I won’t return to this question in the near future, but will simply accept it as a fact and feature
And in my free time I’ll come back and think about whether the ability to separate validation is needed
<img width="1552" alt="Screenshot_2024-06-14_at_4 30 30_AM" src="https://github.com/filamentphp/filament/assets/117383034/43994ea5-2290-4f9b-b91f-c22652d5bc3e">

https://youtu.be/GGwATl3ZSQY
https://youtu.be/XpHHJ2bQHNY
here's the reasons to think that the browser determines the mime type by the file extension

**Comment by cododel** [comment]
I propose to implement the above mentioned acceptedBrowserFileTypes method, which will take precedence for filepond
I've already looked and generally know how to implement it

```php
FileUpload::make('file')
->acceptedBrowserFileTypes(['text/javascript'])
->acceptedFileTypes(['text/plain'])
->getUploadedFileNameForStorageUsing(
fn(TemporaryUploadedFile $file): string => (string) Str::uuid() . '.' . Arr::last(explode('.', $file->getClientOriginalName())),
);
```

**Comment by arc-muns** [comment]
This is still an issue, no reply from maintainers?

**Reply by cododel** [reply]
nothing
